ocaml/typing/typecore.ml

```diff
@@ -370,18 +370,37 @@ let position_and_mode_default = {
   region_mode = None;
 }
 
-(* Calls to functions that have a dot (e.g. [M.f]) are inferred to be nontail. *)
-let should_infer_nontail maybe_ident =
-  match maybe_ident with
-  | Some ident -> begin match ident with
-    | Longident.Ldot _ -> true
-    | _ -> false
+type callee_expr =
+  | Not_a_function_callee
+  | Function_callee of Parsetree.expression
+
+type should_infer_tail_t = Infer_tail | Don't_infer_tail
```

Comment:
I appreciate the wariness towards polymorphic variants you express. In this case, however, I think they would be acceptable, since this type is used only within the body of `should_infer_tail`. If you were returning `should_infer_tail_t` instead of a bool, I'd likely encourage keeping the variant: but the name of the function is clear enough that I prefer keeping the bool return and changing to polymorphic variants instead.